### ISSUE TITLE
kubectl fails to connect to apiserver inside a pod

### ISSUE THREAD

**Issue by a-palchikov** [kubernetes/kubernetes#33019]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
kubectl,defaultconfig
Merge that possibly introduced the error: https://github.com/kubernetes/kubernetes/pull/32613
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Bug report

**Kubernetes version** (use `kubectl version`): v1.4.0-beta.8
```Client Version: version.Info{Major:"1", Minor:"4+", GitVersion:"v1.4.0-beta.8", GitCommit:"3040f87c570a772ce94349b379f41f329494a4f7", GitTreeState:"clean", BuildDate:"2016-09-19T07:54:34Z", GoVersion:"go1.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4+", GitVersion:"v1.4.0-beta.8", GitCommit:"3040f87c570a772ce94349b379f41f329494a4f7", GitTreeState:"clean", BuildDate:"2016-09-19T07:54:34Z", GoVersion:"go1.7", Compiler:"gc", Platform:"linux/amd64"}```

**Environment**:
- **Cloud provider or hardware configuration**: virsh
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 8 (jessie)
- **Kernel** (e.g. `uname -a`): Linux 299e561bce9845aeb537a4a1a50a4562.dev.local-planet 3.10.0-327.el7.x86_64 #1 SMP Thu Oct 29 17:29:29 EDT 2015 x86_64 GNU/Linux
- **Install tools**:
- **Others**:


**What happened**:
kubectl w/o cluster configuration does not connect to apiserver with the following error: `The connection to the server localhost:8080 was refused - did you specify the right host or port?`

**What you expected to happen**:
kubectl w/o parameters used to connect using in-cluster approach, using a token and `KUBERNETES_SERVICE_HOST` / `KUBERNETES_SERVICE_PORT` pair

**How to reproduce it** (as minimally and precisely as possible):
Install a cluster with `k8s 1.4-beta.8`, run a pod with `kubectl` and issue any command

**Anything else do we need to know**:
I think the problem is in this line:
https://github.com/kubernetes/kubernetes/pull/32613/files#diff-47197a425491d7fcfa69115333533db9R116
```go
if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
```
where `err` (apparently) is always `nil`, and (I guess) should be the following instead:
```go
if IsConfigurationInvalid(defErr) && !IsEmptyConfig(defErr) {
```

**Comment by janetkuo** [comment]
cc @madhusudancs

**Comment by janetkuo** [comment]
The fix should probably be cherrypicked to 1.4, cc @pwittrock 

Thanks @a-palchikov for spotting this!

**Comment by madhusudancs** [comment]
@a-palchikov 

I will take a look. But in the meantime:

> Anything else do we need to know:
> I think the problem is in this line:
> https://github.com/kubernetes/kubernetes/pull/32613/files#diff-47197a425491d7fcfa69115333533db9R116

```golang
if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {
```

is right. It shouldn't be `!IsEmptyConfig(defErr)`. See the discussion in your link for details.

**Comment by a-palchikov** [comment]
I see - I'll have to read it more carefully then

**Comment by janetkuo** [comment]
@a-palchikov how did you run a pod with `kubectl`? Did you build your own `kubectl` image? Which `kubectl` version is that image?

**Comment by pwittrock** [comment]
@janetkuo Are you able to reproduce this?

**Comment by pwittrock** [comment]
Looking more closely, it looks like the kubectl version is the beta.8 version.  Did you mount this from the host volume or build it into an image?  What is the best way to reproduce?  I am assuming it is something like:

- untar the beta.8 tar
- run kube-up.sh
- execute `kubect run -t -i --image busybox test`
- in container call `kubectl get pods`

**Comment by a-palchikov** [comment]
@janetkuo I run a pod with `kubectl` mounted from host - it is the same version as the installed k8s (1.4.0-beta.8).
@pwittrock yes, once I've set up k8s, I ran a pod with kubectl mounted from host:
```
kubectl run -i --rm --tty ubuntu --overrides='
{
  "apiVersion": "v1",
  "kind": "Pod",
  "spec": {
    "containers": [
      {
        "name": "curl",
        "image": "tutum/curl:trusty",
        "args": [
          "bash"
        ],
        "stdin": true,
        "stdinOnce": true,
        "tty": true,
        "volumeMounts": [{
          "mountPath": "/usr/bin/kubectl",
          "name": "kubectl"
        }]
      }
    ],
    "volumes": [{
      "name":"kubectl",
      "hostPath":{"path": "/usr/bin/kubectl"}
    }]
  }
}
'  --image=tutum/curl:trusty --restart=Never -- bash
```

**Comment by pwittrock** [comment]
@a-palchikov  Thank you.  That is very helpful.

**Comment by pwittrock** [comment]
@a-palchikov  Thank you.  That is very helpful.

**Comment by janetkuo** [comment]
@caesarxuchao

**Comment by caesarxuchao** [comment]
> `if IsConfigurationInvalid(defErr) && !IsEmptyConfig(err) {`

FWIW, I reproduced the bug, the 1.4 kubectl entered this `if` clause, while I expected it not to. I expected it to load the local config.

I'll read the previous discussion.

**Comment by janetkuo** [comment]
Also, 1.3 kubectl against 1.4 cluster would work

**Comment by caesarxuchao** [comment]
```
	mergedConfig, err := mergedClientConfig.ClientConfig()
	fmt.Println("CHAO: err=", err)
	fmt.Printf("CHAO: mergedConfig=%#v\n", mergedConfig)
```

```
CHAO: err= <nil>
CHAO: mergedConfig=&restclient.Config{Host:"http://localhost:8080", APIPath:"", Prefix:"", ContentConfig:restclient.ContentConfig{AcceptContentTypes:"", ContentType:"", GroupVersion:(*unversioned.GroupVersion)(nil), NegotiatedSerializer:runtime.NegotiatedSerializer(nil)}, Username:"", Password:"", BearerToken:"", Impersonate:"", AuthProvider:(*api.AuthProviderConfig)(nil), AuthConfigPersister:restclient.AuthProviderConfigPersister(nil), TLSClientConfig:restclient.TLSClientConfig{CertFile:"", KeyFile:"", CAFile:"", CertData:[]uint8(nil), KeyData:[]uint8(nil), CAData:[]uint8(nil)}, Insecure:false, UserAgent:"", Transport:http.RoundTripper(nil), WrapTransport:(func(http.RoundTripper) http.RoundTripper)(nil), QPS:0, Burst:0, RateLimiter:flowcontrol.RateLimiter(nil)}
```

**Comment by caesarxuchao** [comment]
https://github.com/kubernetes/kubernetes/blame/master/pkg/kubectl/cmd/util/factory.go#L1168

```
mergo.Merge(&overrides.ClusterDefaults, clientcmd.DefaultCluster)
```

It seems this line sets the `Server`, makes the config non-empty

**Comment by janetkuo** [comment]
@smarterclayton @deads2k

**Comment by caesarxuchao** [comment]
I'm pretty sure this [line](https://github.com/kubernetes/kubernetes/blame/master/pkg/kubectl/cmd/util/factory.go#L1168) is the cause, after removing that line, kubectl worked inside a pod.

This seems to be the bug: `clientcmd.DefaultCluster` is non-empty, it makes overrides non-empty, and it make the mergedConfig [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/merged_client_builder.go#L106) non-empty, so the in-cluster config is not used. 

@smarterclayton could you take a look? It's your https://github.com/kubernetes/kubernetes/pull/30808 introduced the change.

**Comment by janetkuo** [comment]
I can also confirm that kubectl inside a pod worked before #30808

**Comment by smarterclayton** [comment]
The line moved to the factory for kubectl.

**Comment by smarterclayton** [comment]
This is a dupe of another issue - one second.

**Comment by smarterclayton** [comment]
A non-empty default config should be compared against the returned config.  If non-empty default config == returned config, we should try the in-cluster config.

**Comment by janetkuo** [comment]
@smarterclayton do you have a fix for this? We should cherrypick it.

**Comment by caesarxuchao** [comment]
So I think we just need to change the order of the two if clauses here? https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/merged_client_builder.go#L117-L124

**Comment by caesarxuchao** [comment]
No, sorry. @madhusudancs, Maybe just add the `!DeepEqual` as a condition in this if clause: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/clientcmd/merged_client_builder.go#L118?

**Comment by smarterclayton** [comment]
Put up an alternate fix based on my comment on the other PR - please try it

**Comment by a-palchikov** [comment]
Thanks - I'm currently trying

**Comment by caesarxuchao** [comment]
It worked for me.

**Comment by a-palchikov** [comment]
Worked for me as well - thanks again.

**Comment by caesarxuchao** [comment]
I can confirm it's solved by #33136. Could someone else double check?

**Comment by caesarxuchao** [comment]
I'm going to close it. Please reopen if it's not fixed.

**Comment by janetkuo** [comment]
> I can confirm it's solved by #33136. Could someone else double check?

Confirmed
